### PR TITLE
Define non-positive temperature behavior

### LINKED ISSUE #684: Setting `temp=0` does not work as expected
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Setting sampling temperature to `0` should produce valid and "predictable" tokens.

# Current Behavior

Setting temperature to `0` causes sampling to fail completely. This is due to `plogits` being scaled by `1.0f/temp` before sampling [here](https://github.com/ggerganov/llama.cpp/blob/d0a7f742e76bb48c0bd852f0b3bf09ec0b75b200/llama.cpp#L1201). I believe a workaround for this would be to make sampling deterministic when `temp==0` by setting `top_p=0.0` and `top_k=1` and setting `temp>0`.

### PR DESCRIPTION
**Pull request by ivanstepanovftw**
Fixes #684.

### CODE REVIEW

**Comment by ivanstepanovftw** [comment]
Original LLaMA's [`generate` defines temperature](https://github.com/facebookresearch/llama/compare/main...shawwn:llama:main#diff-cfa9e089ebf97ee064b754a2fecad8b5c23610a519a6d640710d94f74fbd8300R65) like so:

            if temperature > 0:
                probs = torch.softmax(logits / temperature, dim=-1)
                next_token = sample(probs, top_p=top_p, top_k=top_k)
            else:
                next_token = torch.argmax(logits, dim=-1)

**Comment by prusnak** [comment]
@abetlen can you please also test this one?

**Comment by abetlen** [comment]
@prusnak tested and can confirm this works as well.

**Review by prusnak — APPROVED** [review]

**Review by Fabio3rs — COMMENTED** [review]
This alternative seems to me to harmonize better with the rest of the code
